File: Assignment 1/Solver.java
import java.util.Scanner;

public class Solver {
  public static final int MAX_COUNT = 30;

  public static void main(String args[]) {
    int count = 0, target;
    int[] array, solution;
    int solutionCount;

    Scanner stdin = new Scanner(System.in);

    while (stdin.hasNextLine()) {   
      count = stdin.nextInt();
      if (count > MAX_COUNT) {
        System.err.println("Invalid count: " + count + ".");
      } else {
        array = new int[count];
        for (int i = 0; i < count; i++) {
          array[i] = stdin.nextInt();
        }
        target = stdin.nextInt();

        solution = new int[count];
        reverseSort(array);
        solutionCount = solve(array, solution, target);
    
        if (solutionCount > 0) {
          for (int i = 0; i < solutionCount; i++) {
            System.out.print(solution[i] + " ");
          }
          System.out.println();
        } else {
          System.out.println("Count = " + count + ": no solution.");
        }
      }
      stdin.nextLine(); // get rid of the newline at the end
    }
    
    stdin.close();
    
    System.out.println("\nEnd of processing.");
    System.exit(0);
  }
  
  public static void reverseSort(int[] array) {
    int max, maxPos;
    for (int i = 0; i < array.length-1; i++) {
      max = array[i];
      maxPos = -1;
      for (int j = i+1; j < array.length; j++) {
        if (array[j] > max) {
          max = array[j];
          maxPos = j;
        }
      }
      if (maxPos > 0) {
        array[maxPos] = array[i];
        array[i] = max;
      }
    }
  }
  
  public static int solve(int[] array, int[] solution, int target) {
    return rsolve(array, array.length, solution, 0, 0, target);
  }

  public static int rsolve(int[] array, int unprocessed, int[] solution, int solutionCount, int soFar, int target) {
    int lastPos, result;
    
    if (soFar == target) {
      result = solutionCount;
    } else if (unprocessed == 0) {
      result = 0;
    } else {
      lastPos = unprocessed - 1;
      result = rsolve(array, lastPos, solution, solutionCount, soFar, target);
      if (result == 0 && soFar+array[lastPos] <= target) {
        solution[solutionCount] = array[lastPos];
        result = rsolve(array, unprocessed-1, solution, solutionCount+1, soFar+array[lastPos], target);
      }
    }
    
    return result;
  }
} 

File: Assignment 1/solver.c
#include <stdio.h>
#include <ctype.h>
#include <string.h>

int rsolve(int *array, int unprocessed, int *solution, int solutionCount, int soFar, int target){
    int lastPos, result;

    if(soFar == target){
        result = solutionCount;
    }
    else if(unprocessed == 0){
        result = 0;
    }
    else{
        lastPos = unprocessed - 1;
        result = rsolve(array, lastPos, solution, solutionCount, soFar, target);
        if(result == 0 && soFar+array[lastPos] <= target){
            solution[solutionCount] = array[lastPos];
            result = rsolve(array, unprocessed - 1, solution, solutionCount + 1, soFar + array[lastPos], target);
        }
    }
    return result;
}

int solve(int *array, int *solution, int target, int size){
    return rsolve(array, size, solution, 0, 0, target);
}

void reverseSort(int *array, int size){
    int max, maxPos;
    
    for(int i = 0; i < size - 1; i++){
        max = array[i];
        maxPos = -1;
        for(int j = i + 1; j < size; j++){
            if(array[j] > max){
                max = array[j];
                maxPos = j;
            }
        }
        if(maxPos > 0){
            array[maxPos] = array[i];
            array[i] = max;
        }
    }
}

int main(){
    const int MAX_COUNT = 30;
    int count = 0, target;  
    int solutionCount; 
    
    while(scanf("%d", &count) != EOF){
        if(count > MAX_COUNT){
            printf("Invalid count %d\n", count);
            char tmp = ' ';
            //scanf("%[^\n]%*c", &tmp);       
            while(tmp != '\n' && tmp != EOF){
                //tmp = getchar();
                scanf("%c", &tmp);
            }
        }
        else{
            int array[MAX_COUNT];  
            int keepTrack = 0;
            for(int i = 0; i <= count; i++){
                scanf("%d", &array[i]);
                keepTrack++;
            } 
           
            int solution[MAX_COUNT];
            int size = keepTrack - 1; //except the target
            target = array[keepTrack-1]; 
            reverseSort(array, size);
           
            solutionCount = solve(array, solution, target, size);
            if(solutionCount > 0){
                for(int i = 0; i < solutionCount; i++){
                   printf("%d ", (solution[i]));
                }
                printf("\n");
            }
            else{
                printf("Count = %d: no solution\n", count);
            }
            array[0] = '\0';
        }
    } 
}